test/distributed/fsdp/test_fsdp_backward_prefetch.py

```diff
@@ -0,0 +1,219 @@
+
+            self.assertEqual(
+                len(args), 2, "expect _get_handle_to_prefetch(state, current_handle)"
+            )
+            state = args[0]
+            current_handle = args[1]
+            training_state = _get_training_state(current_handle)
+            if (
+                training_state == HandleTrainingState.BACKWARD_PRE
+                and state.backward_prefetch == BackwardPrefetch.BACKWARD_PRE
+            ) or (
+                training_state == HandleTrainingState.BACKWARD_POST
+                and state.backward_prefetch == BackwardPrefetch.BACKWARD_POST
+            ):
+                nonlocal all_handle_fqns
+                # FQNs prefixed from the root module
+                # state._exec_order_data.param_to_fqn
+                fqns = get_handle_fqns_from_root(state, handle)
+                all_handle_fqns.append(fqns)
+            return handle
+
+        # flat params from prefetch handle should match
+        # DECODER_PARAM_FQNS and ENCODER_PARAM_FQNS
+        with patch(
+            "torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch",
+            patched_get_handle_to_prefetch,
+        ):
+            for _ in range(NUM_ITERS):
+                optim.zero_grad()
+                loss = model(src, tgt).sum()
+                loss.backward()
+                optim.step()
+                if backward_prefetch is None:
+                    self.assertEqual(len(all_handle_fqns), 0)
+                    continue
+                elif backward_prefetch == BackwardPrefetch.BACKWARD_PRE:
+                    # state._exec_order_data.handles_post_forward_order
+                    # equals forward order
+                    #     encoder 0...5 -> decoder 0...5 -> root
+                    # pre-backward hook order
+                    #     root -> decoder 5...0 -> encoder 5...0
+                    # prefetch order
+                    #     decoder 5...0 -> encoder 5...0 -> None
+                    # None: when current_handle=encoder 0,
+                    #       handles_post_forward_order returns None
+                    # +1 is for the above None
+                    encoder_begin_index = ENCODER_BEGIN_INDEX_FOR_PRE
+                    self.assertEqual(
+                        len(all_handle_fqns), TOTAL_NUM_PREFETCH_FOR_PRE + 1
+                    )
+                elif backward_prefetch == BackwardPrefetch.BACKWARD_POST:
+                    # state._exec_order_data.handles_post_forward_order
+                    # equals forward order
+                    #     encoder 0...5 -> decoder 0...5 -> root
+                    # post-backward hook (AccumulateGrad) order
+                    #     decoder 5, 4...0 -> encoder 5...0 -> root
+                    # prefetch order
+                    #     decoder 4...0 -> encoder 5...0 -> None -> None
+                    # 1st None: when current_handle=encoder 0,
+                    #           handles_post_forward_order returns None
```

Comment:
nit: should "handles_post_forward_order returns None" be something like "_get_handle_to_prefetch returns None"? (same above)